Add node level cache stats for searchable snapshots

tlrx opened the pull request:
This pull request adds a first set of node-level statistics about the shared cache.

The REST endpoint is: `GET /_searchable_snapshots/cache/stats`

And the returned informations are:
```
{
  "nodes" : {
    "eerrtBMtQEisohZzxBLUSw" : {
      "shared_cache" : {
        "reads" : 6051,
        "bytes_read" : "5.1mb",
        "bytes_read_in_bytes" : 5448829,
        "writes" : 37,
        "bytes_written" : "1.1mb",
        "bytes_written_in_bytes" : 1208320,
        "evictions" : 5,
        "num_regions" : 32,
        "size" : "1mb",
        "size_in_bytes" : 1048576,
        "region_size" : "32kb",
        "region__size_in_bytes" : 32768
      }
    }
  }
}

```

I'm opening this PR as a draft since it is lacking tests but I'd like to get feedback about the API and the stats.

tlrx committed:
Add node level cache stats for searchable snapshots

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

tlrx commented:
ping @ywelsch

ywelsch commented:
Brain dump incoming (happy to discuss):

For the "full_cache", I think that hits and misses are interesting. I would not show evictions as there should in theory be none (except when shard is shut down / cache is explicitly cleared, which is not worth tracking I think as it adds more confusion that there is such a thing as "clearing the cache", more on that below). 
For the "partial_cache", I think that hits, misses, and evictions are critical metrics to track. Total regions (and region size) is interesting as well, free and used regions are not (as that number will just go up to max, and then stay at max).

I'm wondering if we should take the opportunity now to rethink a bit on how we present these two "caches". CacheService is more of a shard-level component (except for the "persistence" bit, there is not much use in having it operate at node-level). This could then be renamed to "shard cache" or, even better to remove cache, "shard storage". We could also think about folding the stats for that into shard-level searchable snapshot stats. In contrast, the shared cache is a true node-level cache (and the only one we would take about as "cache") that is evicting parts of one shard for others parts of the same or another shard, and could either be in this separate API here, or it could be folded into node stats proper.

More on naming: I think that "persistent_cache_docs", "persistent_cache_size_in_bytes" etc are too technical terms. This could be "num_files", "total_size_in_bytes" etc. They make more sense at the shard level.

tlrx committed:
Merge branch 'master' into node-level-cache-stats

tlrx committed:
apply feedback (remove full cache stats, add key metrics for shared cache)

tlrx committed:
Merge branch 'master' into node-level-cache-stats

tlrx committed:
add REST spec and test

tlrx committed:
add action name to non-operator actions

tlrx committed:
Merge branch 'master' into node-level-cache-stats

tlrx commented:
@ywelsch I revisited the API according to our discussion (removed full_cache and persistent_cache, free/used regions and added evictions/reads/writes as defined together).

I suggest to move forward and review this PR; I'll add API documentation and tests in follow ups.

ywelsch commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/cache/TransportSearchableSnapshotsNodeCachesStatsAction.java:
```diff
@@ -0,0 +1,270 @@
+        final FrozenCacheService.Stats frozenCacheStats = frozenCacheService.get() != null ? frozenCacheService.get().getStats() : null;
+        return new NodeCachesStatsResponse(
+            clusterService.localNode(),
+            frozenCacheStats != null ? frozenCacheStats.getNumberOfRegions() : 0,
```
maybe there's a dummy `FrozenCacheService.Stats` object that we can use instead of doing all these null checks here?

tlrx replied:
Sure, I pushed 6e901e8292a70901a2b575a73091815345b3a54e

ywelsch commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/cache/TransportSearchableSnapshotsNodeCachesStatsAction.java:
```diff
@@ -0,0 +1,270 @@
+                    builder.humanReadableField("bytes_written_in_bytes", "bytes_written", ByteSizeValue.ofBytes(bytesWritten));
+                    builder.field("evictions", evictions);
+                    builder.field("num_regions", numRegions);
+                    builder.humanReadableField("size_in_bytes", "size", ByteSizeValue.ofBytes(numRegions * regionSize));
```
let's show the size that the user specified, not the rounded down one

tlrx replied:
Ok, I pushed 7d0c96f1abef13551a8c5051171baa2df844cfc4

ywelsch commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/shared/FrozenCacheService.java:
```diff
@@ -595,6 +609,7 @@ public long physicalEndOffset() {
         public boolean tryEvict() {
             if (refCount() <= 1 && evicted.compareAndSet(false, true)) {
                 logger.trace("evicted {} with channel offset {}", regionKey, physicalStartOffset());
+                evictCount.increment(); // only count evictions that are not related to shard shutdown
```
why not count those?

tlrx replied:
Somehow I had the impression that we considered the evictions from shard removal/relocation as part of the normal lifecycle of shards and not interesting in the context of cache statistics.

I pushed aceb3ac215aefb5470e3772e7b1f696a97d75630 to count all eviction. Let me know what you think.

ywelsch reviewed (commented):
I've left minor comments, o.w. looking good.

tlrx committed:
add empty stats instance

tlrx committed:
use cache size setting

tlrx committed:
count all evictions

tlrx committed:
add docs

tlrx commented:
Thanks @ywelsch ! I updated the code according to your comments. I also added a bit of documentation as a recent communication today reminded us to commit the REST API docs alongside the implementation.

Can you please have another look? Thanks

tlrx committed:
Merge branch 'master' into node-level-cache-stats

tlrx committed:
fix doc

ywelsch reviewed (approved):
LGTM (relevant test failure though)

tlrx committed:
remove double underscore

tlrx committed:
Merge branch 'master' into node-level-cache-stats

henningandersen reviewed (approved):
LGTM.

tlrx commented:
Thanks Yannick & Henning!
